### **Group**: Peter Endes-Nagy, Khawaja Hassan, Shah Ali, Viktoria Konya


# **Task2**

Use the senator speeches in the folder 105-extracted-date located in Inputs and do the following:

i) text preprocessing, removing stopwords, punctuations. Justify your choices.

ii) use cosine similarity to find whose senator's speech is closest to senator Biden. Use sen105kh_fix.csv and/or Wikipedia to validate your findings (i.e., understand if the most similar speeches are senators from the same state and/or party). Describe your findings. 

iv) How do your results change if you apply stemming or lemmatization? In your opinion is it better to apply stemming or lemmatization? Why?

In [1]:
# Import libraries
import pandas as pd
import os
from bs4 import BeautifulSoup
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from sklearn.metrics.pairwise import cosine_similarity

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Change path
path_input = "Inputs/105-extracted-date"

Please make sure you have the right path and you're in the Python-Programming-and-Text-Analysis folder

In [3]:
#my_path = ''
#os.chdir(my_path)
os.getcwd()

'/Users/endesnagypeter/Documents/CEU/Python_text/Python-Programming-and-Text-Analysis'

## 1. Data import

In [4]:
# Read in files and create df
speech = []
filename = []
for element in sorted(os.listdir(path_input)):
    if ".txt" in element:
        file = open(os.path.join(path_input,element) , errors = 'ignore' ).readlines() 
        file = ''.join(file) 
        filename.append(element.replace(".txt", ""))
        speech.append(file)

# Put messages to dataframe
df = pd.DataFrame({'filename':filename,'speech':speech})

# Check
pd.set_option("display.max_columns", None)
df[0:10]

filename                                             speech
0   105-abraham-mi  <DOC>\n<DOCNO>105-abraham-mi-1-19981112</DOCNO...
1     105-akaka-hi  <DOC>\n<DOCNO>105-akaka-hi-1-19981021</DOCNO>\...
2    105-allard-co  <DOC>\n<DOCNO>105-allard-co-1-19981009</DOCNO>...
3  105-ashcroft-mo  <DOC>\n<DOCNO>105-ashcroft-mo-1-19981021</DOCN...
4    105-baucus-mt  <DOC>\n<DOCNO>105-baucus-mt-1-19981021</DOCNO>...
5   105-bennett-ut  <DOC>\n<DOCNO>105-bennett-ut-1-19981112</DOCNO...
6     105-biden-de  <DOC>\n<DOCNO>105-biden-de-1-19981021</DOCNO>\...
7  105-bingaman-nm  <DOC>\n<DOCNO>105-bingaman-nm-1-19981021</DOCN...
8      105-bond-mo  <DOC>\n<DOCNO>105-bond-mo-1-19981012</DOCNO>\n...
9     105-boxer-ca  <DOC>\n<DOCNO>105-boxer-ca-1-19981021</DOCNO>\...

## 2. Data preparation

First, we need to retrieve only the text from the files.

For this we are going to:
* Parse the documents with BeautifulSoup
* Remove the words that contain the filename
* Do additional data cleaning: removeing the \n, removing extra spaces and leading spaces.

In [5]:
# Define function to remove the filenames
def replaceName(document, filename):
    document_clean = re.sub(r"(?:\s){}[^, ]*".format(filename), "", document) 
    return str(document_clean)

In [6]:
my_dict = dict.fromkeys(df['filename'])

for ind in df.index:
    
    # Get filename
    filename = df['filename'][ind]
    
    # Get speech
    speech = df['speech'][ind]
    
    # Parse speech
    soup = BeautifulSoup(speech, 'html')
    
    # Remove \n
    s = soup.text.replace("\n", " ")

    # Remove extra spaces
    s = re.sub("\s\s+", " ", s)
    
    # Remove word that contains filename
    s = replaceName(s, filename)
    
    # Remove leading spaces
    s.strip()
    
    my_dict[filename] = s

In [7]:
# Check one element
my_dict['105-abraham-mi'][:300]

' Mr. ABRAHAM. Mr. President, during debate on final passage of the Omnibus Appropriations bill, in which the American Competitiveness and Workforce Improvement Act was included as Title IV of Subdivision C, I asked unanimous consent to have a number of documents printed in the Record. These included'

In [8]:
# Convert back to df
df = pd.DataFrame.from_dict(my_dict, orient='index').rename_axis('filename').reset_index()
df.rename(columns={ df.columns[1]: "speech" }, inplace = True)
df.head()

filename                                             speech
0   105-abraham-mi   Mr. ABRAHAM. Mr. President, during debate on ...
1     105-akaka-hi   Mr. AKAKA. Mr. President, I am pleased that t...
2    105-allard-co   Mr. ALLARD. Mr. President, I rise to make a f...
3  105-ashcroft-mo   Mr. ASHCROFT. Mr. President, the Senate is no...
4    105-baucus-mt   Mr. BAUCUS. I understand that the House has s...

## 3. Data preprocessing and tokenization, applying TF-IDF vectorizer, calculationg cosine similarity

In the next step the following were done:

* We created **text_preprocesser()** function that removes all punctuations, non-alphabetical characeters, stopwords and words with less than 3 characters.
* We applyed the preprocessor on the speeches.
* Then we applied the **TF-IDF vectorizer** where we choped off the lest frequent terms (that appeared in less than 5% of the documents) and the most important ters (that appeared in more than 95% of the documents).

#### 3.1. Text preprocesser

In [9]:
# Define preprocesser
def text_preprocesser(text):
    
    # Remove not alphanumerical characters
    text = re.sub(r'\W',' ', text)
    
    # Remove not alphanumerical characters
    text = re.sub("[^a-zA-Z]+", " ", text)
    
    # Tokenize
    tokens = word_tokenize(text.lower())
    
    # Remove stopwords
    tokens = [token for token in tokens if token not in stopwords.words('english')]
    
    # Remove words with length less than 3 characters
    tokens = [word for word in tokens if len(word)>=3]
    
    preprocessed_text = ' '.join(tokens)
    
    return preprocessed_text

In [10]:
# Check one element
text_preprocesser(df['speech'].iloc[0])[:300]

'abraham president debate final passage omnibus appropriations bill american competitiveness workforce improvement act included title subdivision asked unanimous consent number documents printed record included two documents received administration negotiations whose inclusion seeking help illuminate'

#### 3.2. Apply text preprocesser on the speeches

As it's a time-consuming process, the result were saved, then read back. Please uncomment the code chunk for testing the code. 

In [11]:
##### Apply to df

#df_preprocessed = df['speech'].apply(text_preprocesser)

#### Save

#import pickle
#with open(f'df_preprocessed.p', 'wb') as f:
        #pickle.dump(df_preprocessed, f)

In [12]:
# Import csv again
import pickle
with open(f'df_preprocessed.p', 'rb') as data:
    df_preprocessed = pickle.load(data)

df_preprocessed[0][:300]

'abraham president debate final passage omnibus appropriations bill american competitiveness workforce improvement act included title subdivision asked unanimous consent number documents printed record included two documents received administration negotiations whose inclusion seeking help illuminate'

In [13]:
len(df_preprocessed)

100

#### 3.3. Apply TF-IDF vectorizer

We will:
* ignore terms that appear in less than 5% of the documents and
* ignore terms that appear in more than 95% of the documents

In [14]:
# Apply TF-IDF vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(min_df=0.05, max_df= 0.95) # preprocessing was done before
tfidf = tfidf_vectorizer.fit_transform(df_preprocessed)

In [15]:
# Convert to df
df_tfidf = pd.DataFrame(tfidf.toarray().transpose(), index=tfidf_vectorizer.get_feature_names())
df_tfidf.head(10)

0         1         2         3         4         5   \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aba          0.004530  0.000000  0.000000  0.004474  0.000000  0.000000   
aback        0.000000  0.000000  0.000000  0.001491  0.000000  0.000000   
abandon      0.003513  0.000000  0.001624  0.002313  0.002179  0.003345   
abandoned    0.012813  0.006581  0.006771  0.003013  0.000000  0.000000   
abandoning   0.000000  0.000000  0.000000  0.000000  0.001999  0.000000   
abandonment  0.004225  0.000000  0.002605  0.000000  0.000000  0.000000   

                   6         7         8         9         10        11  \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.002094  0.000000  0.000000   
aba          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
abandon      0.002944  0.003983  0.002058  0.004936  0.004100  0.004691   
abandoned    0.001227  0.004151  0.000000  0.007202  0.002136  0.006518   
abandoning   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
abandonment  0.000944  0.002129  0.000000  0.001583  0.000000  0.010030   

                   12        13        14        15        16        17  \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.000000  0.002205  0.000000   
aba          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
abandon      0.001275  0.000922  0.005709  0.002452  0.002079  0.002438   
abandoned    0.006642  0.024972  0.000000  0.002555  0.000000  0.011854   
abandoning   0.000000  0.000000  0.000000  0.002250  0.000000  0.000000   
abandonment  0.000000  0.001478  0.000000  0.000000  0.000000  0.003909   

                   18        19        20        21        22        23  \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.009087  0.002231  0.000000  0.000000  0.001944  0.000000   
aarp         0.000000  0.033106  0.000000  0.000000  0.000000  0.000000   
aba          0.000000  0.000000  0.000000  0.000000  0.010020  0.002260   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.004521   
abandon      0.002270  0.005573  0.001485  0.003663  0.000000  0.000000   
abandoned    0.002366  0.002323  0.001548  0.000000  0.002024  0.001827   
abandoning   0.000000  0.002046  0.000000  0.000000  0.000000  0.001608   
abandonment  0.003641  0.001788  0.000000  0.002937  0.000000  0.000000   

                   24        25        26        27        28        29  \
aaa          0.000000  0.000000  0.027523  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.001355  0.000000  0.001732  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.000000  0.000000  0.002536   
aba          0.000000  0.000000  0.000000  0.000000  0.004726  0.000000   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
abandon      0.003283  0.001354  0.006934  0.000000  0.003665  0.000399   
abandoned    0.004562  0.004234  0.000000  0.015335  0.000000  0.003738   
abandoning   0.000000  0.000000  0.00000

In [16]:
# Column names
senator_names = df['filename'].str.split('-').str[1] + '-' + df['filename'].str.split('-').str[2] 
senator_names.head()

0     abraham-mi
1       akaka-hi
2      allard-co
3    ashcroft-mo
4      baucus-mt
Name: filename, dtype: object

In [17]:
# Assign column names to matrix
df_tfidf.columns = senator_names
df_tfidf.head(10)

filename     abraham-mi  akaka-hi  allard-co  ashcroft-mo  baucus-mt  \
aaa            0.000000  0.000000   0.000000     0.000000   0.000000   
aapcc          0.000000  0.000000   0.000000     0.000000   0.000000   
aaron          0.000000  0.000000   0.000000     0.000000   0.000000   
aarp           0.000000  0.000000   0.000000     0.000000   0.000000   
aba            0.004530  0.000000   0.000000     0.004474   0.000000   
aback          0.000000  0.000000   0.000000     0.001491   0.000000   
abandon        0.003513  0.000000   0.001624     0.002313   0.002179   
abandoned      0.012813  0.006581   0.006771     0.003013   0.000000   
abandoning     0.000000  0.000000   0.000000     0.000000   0.001999   
abandonment    0.004225  0.000000   0.002605     0.000000   0.000000   

filename     bennett-ut  biden-de  bingaman-nm   bond-mo  boxer-ca  breaux-la  \
aaa            0.000000  0.000000     0.000000  0.000000  0.000000   0.000000   
aapcc          0.000000  0.000000     0.000000  0.000000  0.000000   0.000000   
aaron          0.000000  0.000000     0.000000  0.000000  0.000000   0.000000   
aarp           0.000000  0.000000     0.000000  0.000000  0.002094   0.000000   
aba            0.000000  0.000000     0.000000  0.000000  0.000000   0.000000   
aback          0.000000  0.000000     0.000000  0.000000  0.000000   0.000000   
abandon        0.003345  0.002944     0.003983  0.002058  0.004936   0.004100   
abandoned      0.000000  0.001227     0.004151  0.000000  0.007202   0.002136   
abandoning     0.000000  0.000000     0.000000  0.000000  0.000000   0.000000   
abandonment    0.000000  0.000944     0.002129  0.000000  0.001583   0.000000   

filename     brownback-ks  bryan-nv  bumpers-ar  burns-mt   byrd-wv  \
aaa              0.000000  0.000000    0.000000  0.000000  0.000000   
aapcc            0.000000  0.000000    0.000000  0.000000  0.000000   
aaron            0.000000  0.000000    0.000000  0.000000  0.000000   
aarp             0.000000  0.000000    0.000000  0.000000  0.000000   
aba              0.000000  0.000000    0.000000  0.000000  0.000000   
aback            0.000000  0.000000    0.000000  0.000000  0.000000   
abandon          0.004691  0.001275    0.000922  0.005709  0.002452   
abandoned        0.006518  0.006642    0.024972  0.000000  0.002555   
abandoning       0.000000  0.000000    0.000000  0.000000  0.002250   
abandonment      0.010030  0.000000    0.001478  0.000000  0.000000   

filename     campbell-co  chafee-ri  cleland-ga  coats-in  cochran-ms  \
aaa             0.000000   0.000000    0.000000  0.000000    0.000000   
aapcc           0.000000   0.000000    0.000000  0.000000    0.000000   
aaron           0.000000   0.000000    0.009087  0.002231    0.000000   
aarp            0.002205   0.000000    0.000000  0.033106    0.000000   
aba             0.000000   0.000000    0.000000  0.000000    0.000000   
aback           0.000000   0.000000    0.000000  0.000000    0.000000   
abandon         0.002079   0.002438    0.002270  0.005573    0.001485   
abandoned       0.000000   0.011854    0.002366  0.002323    0.001548   
abandoning      0.000000   0.000000    0.000000  0.002046    0.000000   
abandonment     0.000000   0.003909    0.003641  0.001788    0.000000   

filename     collins-me  conrad-nd  coverdell-ga  craig-id  damato-ny  \
aaa            0.000000   0.000000      0.000000  0.000000   0.000000   
aapcc          0.000000   0.000000      0.000000  0.000000   0.000000   
aaron          0.000000   0.001944      0.000000  0.000000   0.001355   
aarp           0.000000   0.000000      0.000000  0.000000   0.000000   
aba            0.000000   0.010020      0.002260  0.000000   0.000000   
aback          0.000000   0.000000      0.004521  0.000000   0.000000   
abandon        0.003663   0.000000      0.000000  0.003283   0.001354   
abandoned      0.000000   0.002024      0.001827  0.004562   0.004234   
abandoning     0.000000   0.000000      0.001608  0.000000   0.000000   

In [18]:
len(df_tfidf)

22530

#### 4. Apply cosine similarity

In [19]:
from sklearn.metrics.pairwise import cosine_similarity

In [20]:
# Get number of columns
num_col = len(df_tfidf.columns)
num_col

100

In [21]:
# Reshape 
for i in range(1, num_col+1):
    globals()["txt" + str(i)] = df_tfidf[df_tfidf.columns[i-1]].values.reshape(1, -1) 

In [22]:
# Biden
txt7

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.00168191]])

In [23]:
# Similarity between Biden and Roberts - sample
print("Similarity Biden and Roberts:", cosine_similarity(txt7, txt82))

Similarity Biden and Roberts: [[0.39314126]]


In [24]:
# Reshape to vector
import numpy as np
for i in range(1, num_col+1):
    globals()["txt" + str(i)] = np.squeeze(np.asarray(globals()["txt" + str(i)]))

In [25]:
txt7

array([0.        , 0.        , 0.        , ..., 0.        , 0.        ,
       0.00168191])

In [26]:
# Create list of vectors
list_of_txts = []
for i in range(1, num_col+1):
    element = "txt" + str(i)
    list_of_txts.append(element)
print(list_of_txts)

['txt1', 'txt2', 'txt3', 'txt4', 'txt5', 'txt6', 'txt7', 'txt8', 'txt9', 'txt10', 'txt11', 'txt12', 'txt13', 'txt14', 'txt15', 'txt16', 'txt17', 'txt18', 'txt19', 'txt20', 'txt21', 'txt22', 'txt23', 'txt24', 'txt25', 'txt26', 'txt27', 'txt28', 'txt29', 'txt30', 'txt31', 'txt32', 'txt33', 'txt34', 'txt35', 'txt36', 'txt37', 'txt38', 'txt39', 'txt40', 'txt41', 'txt42', 'txt43', 'txt44', 'txt45', 'txt46', 'txt47', 'txt48', 'txt49', 'txt50', 'txt51', 'txt52', 'txt53', 'txt54', 'txt55', 'txt56', 'txt57', 'txt58', 'txt59', 'txt60', 'txt61', 'txt62', 'txt63', 'txt64', 'txt65', 'txt66', 'txt67', 'txt68', 'txt69', 'txt70', 'txt71', 'txt72', 'txt73', 'txt74', 'txt75', 'txt76', 'txt77', 'txt78', 'txt79', 'txt80', 'txt81', 'txt82', 'txt83', 'txt84', 'txt85', 'txt86', 'txt87', 'txt88', 'txt89', 'txt90', 'txt91', 'txt92', 'txt93', 'txt94', 'txt95', 'txt96', 'txt97', 'txt98', 'txt99', 'txt100']


In [27]:
# Create similarity matrix
from scipy import sparse
A =  np.array(np.array([eval(name) for name in list_of_txts]))
A_sparse = sparse.csr_matrix(A)
similarities = cosine_similarity(A_sparse)
print('pairwise dense output:\n {}\n'.format(similarities))

pairwise dense output:
 [[1.         0.11286872 0.1361684  ... 0.09720429 0.16259734 0.13949189]
 [0.11286872 1.         0.07545337 ... 0.04603068 0.07772243 0.06784303]
 [0.1361684  0.07545337 1.         ... 0.0607008  0.09026423 0.08167521]
 ...
 [0.09720429 0.04603068 0.0607008  ... 1.         0.08235814 0.08108126]
 [0.16259734 0.07772243 0.09026423 ... 0.08235814 1.         0.13055899]
 [0.13949189 0.06784303 0.08167521 ... 0.08108126 0.13055899 1.        ]]



In [28]:
# Define similarity matrix
df_similarity = pd.DataFrame(similarities)
df_similarity.columns = senator_names
df_similarity.insert(0,'senator',senator_names)
df_similarity.head()

filename      senator  abraham-mi  akaka-hi  allard-co  ashcroft-mo  \
0          abraham-mi    1.000000  0.112869   0.136168     0.222770   
1            akaka-hi    0.112869  1.000000   0.075453     0.078826   
2           allard-co    0.136168  0.075453   1.000000     0.096285   
3         ashcroft-mo    0.222770  0.078826   0.096285     1.000000   
4           baucus-mt    0.134141  0.109133   0.089875     0.120479   

filename  baucus-mt  bennett-ut  biden-de  bingaman-nm   bond-mo  boxer-ca  \
0          0.134141    0.123907  0.158303     0.143244  0.127530  0.162881   
1          0.109133    0.073202  0.064699     0.108374  0.070092  0.101515   
2          0.089875    0.097429  0.110033     0.116389  0.087289  0.086955   
3          0.120479    0.120272  0.226275     0.180542  0.159450  0.160070   
4          1.000000    0.094567  0.084955     0.110201  0.118031  0.112623   

filename  breaux-la  brownback-ks  bryan-nv  bumpers-ar  burns-mt   byrd-wv  \
0          0.106249      0.140346  0.108282    0.126602  0.123742  0.151738   
1          0.079190      0.051575  0.070959    0.090988  0.085608  0.080040   
2          0.066927      0.076150  0.078450    0.090597  0.084480  0.099059   
3          0.090807      0.185773  0.085601    0.111861  0.142458  0.149835   
4          0.092247      0.068766  0.107518    0.143470  0.322914  0.176853   

filename  campbell-co  chafee-ri  cleland-ga  coats-in  cochran-ms  \
0            0.078532   0.125201    0.135361  0.175003    0.087042   
1            0.050489   0.102484    0.084109  0.070745    0.056415   
2            0.162403   0.089549    0.092877  0.092678    0.077663   
3            0.080298   0.102020    0.141938  0.187907    0.091325   
4            0.080644   0.232519    0.078913  0.078725    0.084261   

filename  collins-me  conrad-nd  coverdell-ga  craig-id  damato-ny  \
0           0.126784   0.117626      0.158389  0.160310   0.118495   
1           0.055630   0.068405      0.044922  0.104363   0.060923   
2           0.067892   0.086660      0.075839  0.124364   0.085107   
3           0.115177   0.135818      0.152225  0.185826   0.101243   
4           0.084905   0.117779      0.060271  0.149323   0.074498   

filename  daschle-sd  dewine-oh   dodd-ct  domenici-nm  dorgan-nd  durbin-il  \
0           0.210736   0.176772  0.193032     0.102856   0.152840   0.185463   
1           0.137096   0.073775  0.091576     0.052401   0.080564   0.078229   
2           0.137095   0.078923  0.105982     0.084675   0.111564   0.095517   
3           0.242532   0.183147  0.192048     0.108019   0.163379   0.185337   
4           0.189076   0.078487  0.121661     0.093363   0.174252   0.103917   

filename   enzi-wy  faircloth-nc  feingold-wi  feinstein-ca   ford-ky  \
0         0.128237      0.145610     0.181447      0.218029  0.075938   
1         0.063998      0.074877     0.085277      0.110003  0.098905   
2         0.131119      0.099127     0.116429      0.099702  0.048927   
3         0.171343      0.139297     0.145098      0.181387  0.073602   
4         0.103620      0.085093     0.116304      0.108637  0.073479   

filename  frist-tn  glenn-oh  gorton-wa  graham-fl  gramm-tx  grams-mn  \
0         0.165500  0.156500   0.142731   0.148300  0.146812  0.179198   
1         0.079206  0.112095   0.103805   0.093976  0.050097  0.079194   
2         0.081861  0.114084   0.095563   0.090110  0.081970  0.135867   
3         0.215327  0.151447   0.141689   0.125553  0.166707  0.164305   
4         0.087470  0.122969   0.135043   0.127204  0.145033  0.110237   

filename  grassley-ia  gregg-nh  hagel-ne  harkin-ia  hatch-ut  helms-nc  \
0            0.143164  0.110481  0.135706   0.136606  0.207545  0.050482   
1            0.070912  0.044968  0.061148   0.073818  0.090675  0.034393   
2            0.084602  0.070411  0.138622   0.087637  0.115318  0.040959   
3            0.160366  0.117136  0.143710   0.166338  0.251176  0.062968   
4            0.107486  0.078333  0.100248 

In [29]:
# Get most similar senator to Biden
df_similarity_nobiden = df_similarity[df_similarity['senator']!= 'biden-de']
df_similarity_nobiden[df_similarity_nobiden['biden-de'] == df_similarity_nobiden['biden-de'].max()][['senator','biden-de']]

filename     senator  biden-de
81        roberts-ks  0.393141

The most similar senator to Biden based on the speech is: Patrick Roberts, Republican senator of Kansas. The similarity between the speeches of the two senators is:  0.393141.

Let's check the 10 most similar senators to Biden.

In [30]:
df_similarity_nobiden[['senator','biden-de']].sort_values('biden-de', ascending = False)[['senator','biden-de']].head(10)

filename       senator  biden-de
81          roberts-ks  0.393141
66        lieberman-ct  0.326581
89            smith-or  0.278569
68            lugar-in  0.276504
61              kyl-az  0.275706
26          daschle-sd  0.261798
51        hutchison-tx  0.256332
7          bingaman-nm  0.253516
83             roth-de  0.248345
70           mccain-az  0.244909

Based on their Wikipedia, from the 10 senators 7 are Republican and 3 are Democrat:
* Republicans: Roberts, Smith, Lugar, Hutchison, Roth, Mccain, Kyl
* Democrats: Lieberman, Daschle, Bingaman
    
The results seem pretty counter-intuitive but what we are doing is checking out word frequencies here. If both sides are touching same topics of the agenda, then the speeches aren’t that much different. Maybe some particular wordings can be different, but getting lost because of the topic-related words.  
Roberts senator from the state of Kansas might be on the top, because he talks about the same topics as Biden and their use of words aren’t that much different. A sentiment analysis could bring light on the differences by checking negative or positive sentiments attached to the same topic. Or some other approach leaving bag of words behind and care about text structure.  

We can also experiment with being more strict on the TF-IDF vectorizer's parameters, excluding only 5-5 percentiles is relatively generous. By increasing the bottom threshold, we implicitly bring the topics closer, by decreasing the upper threshold, we rather exclude same topics. It's a delicate trade-off as the calibration affects other words of interest too, not only the topic-specific ones.  

## 4. Data preprocessing and lemmatization, applying TF-IDF vectorizer, calculaiong cosine similarity

In order to compare our results we chose to use lemmatization instead of using stemming. Lemmatization considers the context and converts the word to its meaningful base form, whereas stemming just removes the last few characters. In out case (speeches), the context and meaning of the words do matter. 

In [31]:
import nltk
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
from nltk.corpus import stopwords
stopwords = nltk.corpus.stopwords.words('english')
from nltk.corpus import wordnet

#### 4.1. Text preprocesser - Lemmatization

In [32]:
# Lemmatizer

def get_wordnet_pos(word):
    """Map POS tag to first character lemmatize() accepts"""
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}

    return tag_dict.get(tag, wordnet.NOUN)

def text_preprocesser_lemmatize(text):
    
    # Remove special characters
    text = re.sub(r'\W',' ', text)
    
    # Remove not alphabet characters
    text = re.sub("[^a-zA-Z]+", " ", text)
    
    # Lowercase and tokenize
    tokens = [word.lower() for word in nltk.word_tokenize(text)]
    
    # Remove stopwords
    tokens = [token for token in tokens if token not in stopwords]
    
    # Remove words with length less than 3 characters
    tokens = [token for token in tokens if len(token)>=3]
    
    lemma = [lemmatizer.lemmatize(token, get_wordnet_pos(token)) for token in tokens]
    
    # Join
    preprocessed_text = ' '.join(lemma)

    return preprocessed_text


In [33]:
# Check one element
text_preprocesser_lemmatize(df['speech'].iloc[0])[:300]

'abraham president debate final passage omnibus appropriation bill american competitiveness workforce improvement act include title subdivision ask unanimous consent number document print record include two document receive administration negotiation whose inclusion seek help illuminate meaning provi'

#### 4.2. Apply text preprocesser

As it's a time-consuming process, the result were saved, then read back. Please uncomment the code chunk for testing the code. 

In [34]:
#### Apply to df

# df_preprocessed_lemmatize = df['speech'].apply(text_preprocesser_lemmatize)

#### Save

# import pickle
# with open(f'df_preprocessed_lemmatize.p', 'wb') as f:
        # pickle.dump(df_preprocessed_lemmatize, f)

In [35]:
with open(f'df_preprocessed_lemmatize.p', 'rb') as data:
    df_preprocessed_lemmatize = pickle.load(data)
df_preprocessed_lemmatize[0][:300]

'abraham president debate final passage omnibus appropriation bill american competitiveness workforce improvement act include title subdivision ask unanimous consent number document print record include two document receive administration negotiation whose inclusion seek help illuminate meaning provi'

In [36]:
len(df_preprocessed_lemmatize)

100

#### 4.3. Apply TF-IDF vectorizer

We will:
* ignore terms that appear in less than 5% of the documents and
* ignore terms that appear in more than 95% of the documents

In [37]:
# Apply TF-IDF vectorizer
tfidf_vectorizer = TfidfVectorizer(min_df=0.05, max_df= 0.95) # preprocessing was done before
tfidf_lemma = tfidf_vectorizer.fit_transform(df_preprocessed_lemmatize)

In [38]:
# Convert to df
df_tfidf_lemma = pd.DataFrame(tfidf_lemma.toarray().transpose(), index=tfidf_vectorizer.get_feature_names())
df_tfidf_lemma.head(10)

0         1         2         3         4         5   \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aba          0.004618  0.000000  0.000000  0.004740  0.000000  0.000000   
aback        0.000000  0.000000  0.000000  0.001580  0.000000  0.000000   
abandon      0.014480  0.005557  0.007081  0.004693  0.002841  0.002961   
abandonment  0.004307  0.000000  0.002668  0.000000  0.000000  0.000000   
abate        0.000000  0.000000  0.000000  0.001204  0.006560  0.000000   
abatement    0.000000  0.000000  0.000000  0.001274  0.000000  0.000000   

                   6         7         8         9        10        11  \
aaa          0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   
aaron        0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   
aarp         0.000000  0.000000  0.000000  0.002121  0.00000  0.000000   
aba          0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   
aback        0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   
abandon      0.003554  0.007270  0.002406  0.010212  0.00527  0.009655   
abandonment  0.000957  0.002283  0.000000  0.001603  0.00000  0.010394   
abate        0.001172  0.000000  0.002777  0.000000  0.00000  0.000000   
abatement    0.000000  0.000000  0.000000  0.000000  0.00000  0.000000   

                   12        13        14        15        16        17  \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.000000  0.002196  0.000000   
aba          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
abandon      0.006660  0.022228  0.005348  0.005962  0.001762  0.011882   
abandonment  0.000000  0.001551  0.000000  0.000000  0.000000  0.003950   
abate        0.002563  0.000000  0.000000  0.001251  0.000000  0.001614   
abatement    0.000000  0.000000  0.000000  0.000000  0.000000  0.003414   

                   18        19        20        21        22        23  \
aaa          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.009305  0.002323  0.000000  0.000000  0.002016  0.000000   
aarp         0.000000  0.034470  0.000000  0.000000  0.000000  0.000000   
aba          0.000000  0.000000  0.000000  0.000000  0.010390  0.002327   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.004653   
abandon      0.003958  0.007904  0.002516  0.003190  0.001715  0.002304   
abandonment  0.003728  0.001861  0.000000  0.003005  0.000000  0.000000   
abate        0.000000  0.000000  0.002904  0.000000  0.000000  0.000000   
abatement    0.000000  0.000000  0.000000  0.000000  0.000000  0.003751   

                   24        25        26        27        28        29  \
aaa          0.000000  0.000000  0.028858  0.000000  0.000000  0.000000   
aapcc        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
aaron        0.000000  0.001377  0.000000  0.001699  0.000000  0.000000   
aarp         0.000000  0.000000  0.000000  0.000000  0.000000  0.002560   
aba          0.000000  0.000000  0.000000  0.000000  0.004861  0.000000   
aback        0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
abandon      0.006654  0.004685  0.006189  0.012288  0.003209  0.003424   
abandonment  0.003582  0.000000  0.000000  0.001362  0.000000  0.000645   
abate        0.000000  0.000000  0.000000  0.000000

In [39]:
# Assign column names to matrix
df_tfidf_lemma.columns = senator_names
df_tfidf_lemma.head(10)

filename     abraham-mi  akaka-hi  allard-co  ashcroft-mo  baucus-mt  \
aaa            0.000000  0.000000   0.000000     0.000000   0.000000   
aapcc          0.000000  0.000000   0.000000     0.000000   0.000000   
aaron          0.000000  0.000000   0.000000     0.000000   0.000000   
aarp           0.000000  0.000000   0.000000     0.000000   0.000000   
aba            0.004618  0.000000   0.000000     0.004740   0.000000   
aback          0.000000  0.000000   0.000000     0.001580   0.000000   
abandon        0.014480  0.005557   0.007081     0.004693   0.002841   
abandonment    0.004307  0.000000   0.002668     0.000000   0.000000   
abate          0.000000  0.000000   0.000000     0.001204   0.006560   
abatement      0.000000  0.000000   0.000000     0.001274   0.000000   

filename     bennett-ut  biden-de  bingaman-nm   bond-mo  boxer-ca  breaux-la  \
aaa            0.000000  0.000000     0.000000  0.000000  0.000000    0.00000   
aapcc          0.000000  0.000000     0.000000  0.000000  0.000000    0.00000   
aaron          0.000000  0.000000     0.000000  0.000000  0.000000    0.00000   
aarp           0.000000  0.000000     0.000000  0.000000  0.002121    0.00000   
aba            0.000000  0.000000     0.000000  0.000000  0.000000    0.00000   
aback          0.000000  0.000000     0.000000  0.000000  0.000000    0.00000   
abandon        0.002961  0.003554     0.007270  0.002406  0.010212    0.00527   
abandonment    0.000000  0.000957     0.002283  0.000000  0.001603    0.00000   
abate          0.000000  0.001172     0.000000  0.002777  0.000000    0.00000   
abatement      0.000000  0.000000     0.000000  0.000000  0.000000    0.00000   

filename     brownback-ks  bryan-nv  bumpers-ar  burns-mt   byrd-wv  \
aaa              0.000000  0.000000    0.000000  0.000000  0.000000   
aapcc            0.000000  0.000000    0.000000  0.000000  0.000000   
aaron            0.000000  0.000000    0.000000  0.000000  0.000000   
aarp             0.000000  0.000000    0.000000  0.000000  0.000000   
aba              0.000000  0.000000    0.000000  0.000000  0.000000   
aback            0.000000  0.000000    0.000000  0.000000  0.000000   
abandon          0.009655  0.006660    0.022228  0.005348  0.005962   
abandonment      0.010394  0.000000    0.001551  0.000000  0.000000   
abate            0.000000  0.002563    0.000000  0.000000  0.001251   
abatement        0.000000  0.000000    0.000000  0.000000  0.000000   

filename     campbell-co  chafee-ri  cleland-ga  coats-in  cochran-ms  \
aaa             0.000000   0.000000    0.000000  0.000000    0.000000   
aapcc           0.000000   0.000000    0.000000  0.000000    0.000000   
aaron           0.000000   0.000000    0.009305  0.002323    0.000000   
aarp            0.002196   0.000000    0.000000  0.034470    0.000000   
aba             0.000000   0.000000    0.000000  0.000000    0.000000   
aback           0.000000   0.000000    0.000000  0.000000    0.000000   
abandon         0.001762   0.011882    0.003958  0.007904    0.002516   
abandonment     0.000000   0.003950    0.003728  0.001861    0.000000   
abate           0.000000   0.001614    0.000000  0.000000    0.002904   
abatement       0.000000   0.003414    0.000000  0.000000    0.000000   

filename     collins-me  conrad-nd  coverdell-ga  craig-id  damato-ny  \
aaa            0.000000   0.000000      0.000000  0.000000   0.000000   
aapcc          0.000000   0.000000      0.000000  0.000000   0.000000   
aaron          0.000000   0.002016      0.000000  0.000000   0.001377   
aarp           0.000000   0.000000      0.000000  0.000000   0.000000   
aba            0.000000   0.010390      0.002327  0.000000   0.000000   
aback          0.000000   0.000000      0.004653  0.000000   0.000000   
abandon        0.003190   0.001715      0.002304  0.006654   0.004685   
abandonment    0.003005   0.000000      0.000000  0.003582   0.000000   
abate          0.000000   0.000000      0.000000  0.000000   0.000000   

Note that the number of tokens decreased by about 6000 terms compared to using the raw tokenized dataset without lemmatization.

In [40]:
len(df_tfidf_lemma) - len(df_tfidf)

-6255

#### 4. Apply cosine similarity

In [41]:
# Reshape 
for i in range(1, num_col+1):
    globals()["txt" + str(i)] = df_tfidf_lemma[df_tfidf_lemma.columns[i-1]].values.reshape(1, -1) 

In [42]:
# Biden
txt7

array([[0.        , 0.        , 0.        , ..., 0.02007284, 0.        ,
        0.00170392]])

In [43]:
# Similarity between Biden and Roberts - sample
print("Similarity Biden and Roberts:", cosine_similarity(txt7, txt82))

Similarity Biden and Roberts: [[0.47130203]]


In [44]:
# Reshape to vector
import numpy as np
for i in range(1, num_col+1):
    globals()["txt" + str(i)] = np.squeeze(np.asarray(globals()["txt" + str(i)]))

In [45]:
txt7

array([0.        , 0.        , 0.        , ..., 0.02007284, 0.        ,
       0.00170392])

In [46]:
# Create list of vectors
list_of_txts = []
for i in range(1, num_col+1):
    element = "txt" + str(i)
    list_of_txts.append(element)
print(list_of_txts)

['txt1', 'txt2', 'txt3', 'txt4', 'txt5', 'txt6', 'txt7', 'txt8', 'txt9', 'txt10', 'txt11', 'txt12', 'txt13', 'txt14', 'txt15', 'txt16', 'txt17', 'txt18', 'txt19', 'txt20', 'txt21', 'txt22', 'txt23', 'txt24', 'txt25', 'txt26', 'txt27', 'txt28', 'txt29', 'txt30', 'txt31', 'txt32', 'txt33', 'txt34', 'txt35', 'txt36', 'txt37', 'txt38', 'txt39', 'txt40', 'txt41', 'txt42', 'txt43', 'txt44', 'txt45', 'txt46', 'txt47', 'txt48', 'txt49', 'txt50', 'txt51', 'txt52', 'txt53', 'txt54', 'txt55', 'txt56', 'txt57', 'txt58', 'txt59', 'txt60', 'txt61', 'txt62', 'txt63', 'txt64', 'txt65', 'txt66', 'txt67', 'txt68', 'txt69', 'txt70', 'txt71', 'txt72', 'txt73', 'txt74', 'txt75', 'txt76', 'txt77', 'txt78', 'txt79', 'txt80', 'txt81', 'txt82', 'txt83', 'txt84', 'txt85', 'txt86', 'txt87', 'txt88', 'txt89', 'txt90', 'txt91', 'txt92', 'txt93', 'txt94', 'txt95', 'txt96', 'txt97', 'txt98', 'txt99', 'txt100']


In [47]:
# Create similarity matrix
from scipy import sparse
A =  np.array(np.array([eval(name) for name in list_of_txts]))
A_sparse = sparse.csr_matrix(A)
similarities = cosine_similarity(A_sparse)
print('pairwise dense output:\n {}\n'.format(similarities))

pairwise dense output:
 [[1.         0.11048625 0.12188617 ... 0.08611824 0.15113374 0.12806296]
 [0.11048625 1.         0.06546759 ... 0.04187035 0.07262217 0.06461585]
 [0.12188617 0.06546759 1.         ... 0.05451684 0.08047845 0.07196163]
 ...
 [0.08611824 0.04187035 0.05451684 ... 1.         0.07501323 0.07624209]
 [0.15113374 0.07262217 0.08047845 ... 0.07501323 1.         0.12278158]
 [0.12806296 0.06461585 0.07196163 ... 0.07624209 0.12278158 1.        ]]



In [48]:
# Define similarity matrix
df_similarity = pd.DataFrame(similarities)
df_similarity.columns = senator_names
df_similarity.insert(0,'senator',senator_names)
df_similarity.head()

filename      senator  abraham-mi  akaka-hi  allard-co  ashcroft-mo  \
0          abraham-mi    1.000000  0.110486   0.121886     0.214059   
1            akaka-hi    0.110486  1.000000   0.065468     0.072664   
2           allard-co    0.121886  0.065468   1.000000     0.092392   
3         ashcroft-mo    0.214059  0.072664   0.092392     1.000000   
4           baucus-mt    0.123617  0.098570   0.078192     0.111123   

filename  baucus-mt  bennett-ut  biden-de  bingaman-nm   bond-mo  boxer-ca  \
0          0.123617    0.110593  0.151989     0.129213  0.150210  0.157791   
1          0.098570    0.064234  0.063162     0.093485  0.086077  0.094349   
2          0.078192    0.080981  0.101882     0.103268  0.108509  0.076696   
3          0.111123    0.104321  0.226768     0.163228  0.195251  0.153874   
4          1.000000    0.081132  0.078265     0.096451  0.126748  0.104981   

filename  breaux-la  brownback-ks  bryan-nv  bumpers-ar  burns-mt   byrd-wv  \
0          0.095588      0.134765  0.105546    0.119687  0.120104  0.141327   
1          0.078915      0.048435  0.070707    0.070815  0.077662  0.077670   
2          0.058329      0.069460  0.075152    0.074783  0.073945  0.086406   
3          0.084892      0.166423  0.079495    0.100717  0.146820  0.140848   
4          0.085839      0.063299  0.101078    0.125296  0.338919  0.162266   

filename  campbell-co  chafee-ri  cleland-ga  coats-in  cochran-ms  \
0            0.072466   0.116841    0.125299  0.176683    0.079305   
1            0.044663   0.101087    0.083577  0.066125    0.052789   
2            0.164342   0.083928    0.084419  0.084638    0.077257   
3            0.077476   0.093370    0.134395  0.181675    0.087047   
4            0.075010   0.228886    0.072356  0.070646    0.082784   

filename  collins-me  conrad-nd  coverdell-ga  craig-id  damato-ny  \
0           0.121118   0.108866      0.136141  0.145726   0.105709   
1           0.051590   0.070232      0.041102  0.101871   0.052757   
2           0.059844   0.083738      0.070937  0.111811   0.075138   
3           0.115632   0.134773      0.135973  0.174711   0.089264   
4           0.079888   0.118288      0.052525  0.139724   0.068561   

filename  daschle-sd  dewine-oh   dodd-ct  domenici-nm  dorgan-nd  durbin-il  \
0           0.199346   0.168373  0.177827     0.091320   0.141605   0.179400   
1           0.133835   0.065665  0.082087     0.046288   0.081157   0.073214   
2           0.133717   0.070808  0.091461     0.076130   0.101601   0.081377   
3           0.234666   0.174219  0.174967     0.097335   0.155381   0.183358   
4           0.183162   0.068685  0.110021     0.081857   0.170368   0.092139   

filename   enzi-wy  faircloth-nc  feingold-wi  feinstein-ca   ford-ky  \
0         0.110488      0.139630     0.166749      0.214007  0.069387   
1         0.057167      0.072905     0.073599      0.105232  0.102397   
2         0.121183      0.091644     0.097902      0.088859  0.042602   
3         0.167183      0.138362     0.134596      0.175226  0.070592   
4         0.091967      0.075711     0.103888      0.099761  0.069881   

filename  frist-tn  glenn-oh  gorton-wa  graham-fl  gramm-tx  grams-mn  \
0         0.152558  0.152071   0.127789   0.138701  0.131590  0.166242   
1         0.068755  0.109340   0.095666   0.088423  0.047279  0.074221   
2         0.070726  0.105416   0.084704   0.082921  0.072132  0.124386   
3         0.209229  0.146728   0.124806   0.117348  0.154125  0.156386   
4         0.077240  0.124119   0.125820   0.121063  0.115120  0.103585   

filename  grassley-ia  gregg-nh  hagel-ne  harkin-ia  hatch-ut  helms-nc  \
0            0.129377  0.099104  0.126492   0.121284  0.196111  0.071877   
1            0.066862  0.039577  0.055898   0.071746  0.083275  0.050783   
2            0.073762  0.063567  0.125935   0.077832  0.101616  0.057995   
3            0.148036  0.111201  0.139797   0.157188  0.261714  0.082818   
4            0.099374  0.070076  0.096443 

In [49]:
# Get most similar senator to Biden
df_similarity_nobiden = df_similarity[df_similarity['senator']!= 'biden-de']
df_similarity_nobiden[df_similarity_nobiden['biden-de'] == df_similarity_nobiden['biden-de'].max()][['senator','biden-de']]

filename     senator  biden-de
81        roberts-ks  0.471302

With the lemmatization the most similar senator to Biden based on the speech is still Patrick Roberts, Republican senator of Kansas. However, the similarity index between the two senators' speeches slightly increased from 0.393141 to 0.471302.

Let's ceheck the 10 most similar senators to Biden.

In [50]:
df_similarity_nobiden[['senator','biden-de']].sort_values('biden-de', ascending = False).head(10)

filename       senator  biden-de
81          roberts-ks  0.471302
66        lieberman-ct  0.326881
89            smith-or  0.277172
68            lugar-in  0.274742
26          daschle-sd  0.260803
61              kyl-az  0.257698
51        hutchison-tx  0.252216
7          bingaman-nm  0.248942
83             roth-de  0.244127
72         mikulski-md  0.238297

The picture is very similar to the simple tokenization case. The difference is that with lemmatization in the top 10 most similar senator we have 6 Republicans and 4 Democrats. Instead of McCain Democrat senator Mikulski appeared rank as the 10th most similar senator to Biden. 